File: python/백준/그래프/BFS/2589 보물섬.py
import sys
from collections import deque

input = sys.stdin.readline

N, M = map(int, input().split())

treasure_map = [list(input().strip()) for _ in range(N)]  # N줄, 각 줄은 문자 리스트

directions = [(0,1), (1,0), (-1,0), (0,-1)]

def BFS(y, x):
    queue = deque([(y,x)])

    visited = [[-1] * M for _ in range(N)]
    visited[y][x] = 0
    max_dist = 0
    while queue:
        y, x = queue.popleft()

        for dy, dx in directions:
            ny, nx = y + dy, x + dx

            if 0 <= ny < N and 0 <= nx < M:
                if  visited[ny][nx] == -1 and treasure_map[ny][nx] == 'L':
                    visited[ny][nx] = visited[y][x] + 1
                    max_dist = max(max_dist, visited[ny][nx])
                    queue.append((ny,nx))
    return max_dist

ans = 0

for i in range(M):
    for j in range(N):
        if treasure_map[j][i] == 'L':
            dist = BFS(j,i)

            ans = max(dist, ans)

print(ans)

File: cpp/백준/그래프/BFS/2589 보물섬.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <string>
#include <algorithm>
using namespace std;

int N, M;
vector<string> treasure_map;

int BFS(int sy, int sx) {
    int dy[] = {0, 1, -1, 0};
    int dx[] = {1, 0, 0, -1};
    vector<vector<int>> visited(N, vector<int>(M, -1));
    queue<pair<int, int>> q;
    q.push({sy, sx});
    visited[sy][sx] = 0;
    int max_dist = 0;

    while (!q.empty()) {
        auto [y, x] = q.front();
        q.pop();
        for (int i = 0; i < 4; ++i) {
            int ny = y + dy[i];
            int nx = x + dx[i];
            if (ny < 0 || ny >= N || nx < 0 || nx >= M) continue;
            if (visited[ny][nx] == -1 && treasure_map[ny][nx] == 'L') {
                visited[ny][nx] = visited[y][x] + 1;
                max_dist = max(max_dist, visited[ny][nx]);
                q.push({ny, nx});
            }
        }
    }
    return max_dist;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> N >> M;
    treasure_map.resize(N);
    for (int i = 0; i < N; ++i) cin >> treasure_map[i];

    int ans = 0;
    for (int i = 0; i < M; ++i) {
        for (int j = 0; j < N; ++j) {
            if (treasure_map[j][i] == 'L') {
                int dist = BFS(j, i);
                ans = max(ans, dist);
            }
        }
    }

    cout << ans << "\n";
    return 0;
}
